python/pdo/contract/transaction.py

```diff
@@ -56,56 +63,69 @@ def __get(self, contractid, statehash) :
+    def FindDependency(self, contractid, statehash) :
         logger.debug('find dependency for %s, %s', contractid, statehash)
 
-        txnid = self.__get(contractid, statehash)
-        if txnid :
-            return txnid
-
-        # no information about this update locally, so go to the
-        # ledger to retrieve it
-        registry_helper = create_submitter(ledger_config)
-
-        try :
-            # this is not very efficient since it pulls all of the state
-            # down with the txnid
-            contract_state_info = registry_helper.get_state_details(contractid, statehash)
-            txnid = contract_state_info['transaction_id']
-            self.__set(contractid, statehash, txnid)
-            return txnid
-        except Exception as e :
-            logger.info('failed to retrieve the transaction: %s', str(e))
-
-        logger.info('unable to find dependency for %s:%s', contractid, statehash)
-        return None
+        with self.__lock__ :
+            txnid = self.__get(contractid, statehash)
+            if txnid :
+                return txnid
+
+            try :
+                # this is not very efficient since it pulls all of the state
+                # down with the txnid
+                contract_state_info = self.__submitter__.get_state_details(contractid, statehash)
+                txnid = contract_state_info['transaction_id']
+                self.__set(contractid, statehash, txnid)
+                return txnid
+            except Exception as e :
+                logger.info('unable to find dependency for %s:%s; failed to retrieve the transaction', contractid, statehash)
+                return None
 
     ## -------------------------------------------------------
     def SaveDependency(self, contractid, statehash, txnid) :
-        self.__set(contractid, statehash, txnid)
+        with self.__lock__ :
+            self.__set(contractid, statehash, txnid)
 
 # -----------------------------------------------------------------
-__dependencies__ = Dependencies()
-__lock_for_dependencies__ = threading.RLock()
+def start_transaction_processing_service(ledger_config = None):
+    global transaction_task_queue
+    with transaction_thread_lock :
+        if transaction_task_queue :
+            return
+        else :
+            transaction_task_queue = queue.Queue()
 
-# -----------------------------------------------------------------
-def start_transaction_processing_service():
+    if ledger_config is None :
+        ledger_config = pconfig.shared_configuration(['Ledger'])
+
+    global registry_helper
+    registry_helper = create_submitter(ledger_config)
+
+    global __dependencies__
+    __dependencies__ = Dependencies(registry_helper)
 
-    __transaction_executor__.submit(__transaction_worker__)
+    logger.debug('start transaction service threads')
+    for i in range(ledger_config.get("transaction_service_threads", 1)) :
+        thread = threading.Thread(target=__transaction_worker__)
```

Comment:
this is interesting, and definitely useful to have more than one transaction threads (I suppose currently only have 1). If I understand correctly, we have a single queue from which all transaction threads get their task, so I suppose no change to transaction_worker() in order to add more transaction threads.

Reply:
we are limited only to one because I wasn't sure if we could add more transaction manager threads. the transaction manager is non-blocking for now... but that is again problematic logic.